File: ABC/351-400/376/d.py
from collections import deque

def main():
  n,m = map(int,input().split())
  g = [[] for _ in range(n)]
  for i in range(m):
    a,b = map(int,input().split())
    a-=1; b-=1
    g[a].append(b)
  INF = 1001001001
  dist = [INF]*n
  dist[0] = 0
  q = deque()
  q.append(0)
  ans = INF
  while(len(q)):
    v = q.popleft()
    for u in g[v]:
      if dist[u] != INF:
        if u == 0:
          ans = min(ans,dist[v]+1)
        continue
      q.append(u)
      dist[u] = dist[v]+1
  if ans == INF:
    ans = -1
  print(ans)
 
  
if __name__ == '__main__':
  main()

File: ABC/351-400/376/d.cpp
#include <bits/stdc++.h>
using namespace std;
#define rep(i,n) for(int i=0;i<n;++i)

int main(){
    int n,m;
    cin >> n >> m;
    vector<vector<int> > g(n);
    vector<pair<int,int> > es;
    rep(i,m){
        int a,b;
        cin >> a >> b;
        a--; b--;
        g[a].push_back(b);
        es.emplace_back(a,b);
    }
    const int INF = 1001001001;
    vector<int> dist(n,INF);
    dist[0] = 0;
    queue<int> q;
    q.push(0);
    while(!q.empty()){
        int v = q.front(); q.pop();
        for(int u : g[v]){
            if(dist[u] != INF) continue;
            dist[u] = dist[v]+1;
            q.push(u);
        }
    }
    int ans = INF;
    rep(i,m){
        auto [a,b] = es[i];
        if(b == 0) ans = min(ans,dist[a]+1);
    }
    if(ans == INF) cout << -1 << endl;
    else cout << ans << endl;
    return 0;
}